Adopt dynamicDowncast<> in editing's HTMLConverter

annevk opened the pull request:
#### 65b42b723b10db32162e2567cd7a54136149a69b
<pre>
Adopt dynamicDowncast&lt;&gt; in editing&apos;s HTMLConverter
<a href="https://bugs.webkit.org/show_bug.cgi?id=268442">https://bugs.webkit.org/show_bug.cgi?id=268442</a>

Reviewed by Chris Dumez.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverterCaches::inlineStylePropertyForElement):
(stringFromCSSValue):
(HTMLConverterCaches::propertyValueForNode):
(HTMLConverterCaches::floatPropertyValueForNode):
(HTMLConverter::_blockLevelElementForNode):
(HTMLConverterCaches::colorPropertyValueForNode):
(HTMLConverter::_addTableCellForElement):
(HTMLConverter::_processElement):
(HTMLConverter::_traverseNode):
(fileWrapperForElement):
(WebCore::editingAttributedString):

Canonical link: <a href="https://commits.webkit.org/273881@main">https://commits.webkit.org/273881@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/381ce987ec2d73abea633e2593a6d302ebcb6734

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36870 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15803 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39137 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39507 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33013 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38155 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18320 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12920 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31567 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37432 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13347 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32577 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11652 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11675 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32869 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40757 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33447 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33211 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37572 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11951 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9760 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35701 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13609 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8374 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12341 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12859 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -480,72 +485,73 @@ static bool stringFromCSSValue(CSSValue& value, String& result)
     switch (propertyId) {
     case CSSPropertyDisplay:
-        if (element.hasTagName(headTag) || element.hasTagName(scriptTag) || element.hasTagName(appletTag) || element.hasTagName(noframesTag))
+        if (element->hasTagName(headTag) || element->hasTagName(scriptTag) || element->hasTagName(appletTag) || element->hasTagName(noframesTag))
```
Should we make this a `switch(element->nodeName())`? Wasn't sure how much additional refactoring I should throw in. Could also do it as a follow-up.

cdumez replied:
Given the number of checks here, I think switching to a `switch(element->nodeName())` is a no-brainer. Feel free to do in this patch or as a follow-up.

cdumez commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -456,21 +461,21 @@ static bool stringFromCSSValue(CSSValue& value, String& result)
     bool inherit = false;
-    Element& element = downcast<Element>(node);
-    if (RefPtr<CSSValue> value = computedStylePropertyForElement(element, propertyId)) {
+    if (RefPtr<CSSValue> value = computedStylePropertyForElement(*element, propertyId)) {
```
Can drop `<CSSValue>`

cdumez commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -456,21 +461,21 @@ static bool stringFromCSSValue(CSSValue& value, String& result)
     }
 
-    if (RefPtr<CSSValue> value = inlineStylePropertyForElement(element, propertyId)) {
+    if (RefPtr<CSSValue> value = inlineStylePropertyForElement(*element, propertyId)) {
```
Can drop `<CSSValue>`

cdumez commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -604,21 +610,21 @@ static inline bool floatValueFromPrimitiveValue(CSSPrimitiveValue& primitiveValu
-        if (ContainerNode* parent = node.parentInComposedTree())
+    RefPtr element = dynamicDowncast<Element>(node);
+    if (!element) {
+        if (auto* parent = node.parentInComposedTree())
```
We likely want to ref the parent here.

cdumez commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -638,7 +644,7 @@ static inline bool floatValueFromPrimitiveValue(CSSPrimitiveValue& primitiveValu
 
     if (inherit) {
-        if (ContainerNode* parent = node.parentInComposedTree())
+        if (auto* parent = node.parentInComposedTree())
```
We likely want to ref the parent here.

cdumez commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -721,7 +727,9 @@ static inline bool floatValueFromPrimitiveValue(CSSPrimitiveValue& primitiveValu
 Element* HTMLConverter::_blockLevelElementForNode(Node* node)
 {
-    Element* element = is<Element>(node) ? downcast<Element>(node) : node->parentElement();
+    auto* element = dynamicDowncast<Element>(node);
```
We should probably ref the element here.

cdumez commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -744,22 +752,22 @@ static Color normalizedColor(Color color, bool ignoreDefaultColor, Element& elem
-        if (Node* parent = node.parentInComposedTree())
+    RefPtr element = dynamicDowncast<Element>(node);
+    if (!element) {
+        if (auto* parent = node.parentInComposedTree())
```
RefPtr

cdumez commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -781,7 +789,7 @@ static Color normalizedColor(Color color, bool ignoreDefaultColor, Element& elem
 
     if (inherit) {
-        if (Node* parent = node.parentInComposedTree())
+        if (auto* parent = node.parentInComposedTree())
```
RefPtr

cdumez commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1677,10 +1683,9 @@ static NSInteger _colCompare(id block1, id block2, void *)
-            NSURL *url = [NSURL fileURLWithPath:attachment.file()->path()];
+    } else if (RefPtr attachment = dynamicDowncast<HTMLAttachmentElement>(element)) {
+        if (attachment->file()) {
+            NSURL *url = [NSURL fileURLWithPath:attachment->file()->path()];
```
Would be good to use RetainPtr here.

cdumez commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1722,8 +1727,8 @@ static NSInteger _colCompare(id block1, id block2, void *)
-    } else if (is<HTMLFrameElementBase>(element)) {
-        if (Document* contentDocument = downcast<HTMLFrameElementBase>(element).contentDocument()) {
+    } else if (auto* frameElement = dynamicDowncast<HTMLFrameElementBase>(element)) {
+        if (auto* contentDocument = frameElement->contentDocument()) {
```
RefPtr

cdumez commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1754,26 +1759,24 @@ static NSInteger _colCompare(id block1, id block2, void *)
-                NSString *value = inputElement.value();
+        if (RefPtr inputElement = dynamicDowncast<HTMLInputElement>(element)) {
+            if (inputElement->type() == textAtom()) {
+                NSString *value = inputElement->value();
```
RetainPtr

cdumez commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1754,26 +1759,24 @@ static NSInteger _colCompare(id block1, id block2, void *)
-            HTMLTextAreaElement& textAreaElement = downcast<HTMLTextAreaElement>(element);
-            NSString *value = textAreaElement.value();
+        if (RefPtr textAreaElement = dynamicDowncast<HTMLTextAreaElement>(element)) {
+            NSString *value = textAreaElement->value();
```
RetainPtr

cdumez commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2232,8 +2234,8 @@ static NSInteger _colCompare(id block1, id block2, void *)
-    if (is<RenderImage>(renderer)) {
-        auto* image = downcast<RenderImage>(*renderer).cachedImage();
+    if (auto* renderImage = dynamicDowncast<RenderImage>(renderer)) {
+        auto* image = renderImage->cachedImage();
```
`CachedResourceHandle`

cdumez reviewed (approved):
LGTM with nits.

cdumez commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1677,12 +1730,11 @@ static NSInteger _colCompare(id block1, id block2, void *)
-            NSURL *url = [NSURL fileURLWithPath:attachment.file()->path()];
+    } else if (RefPtr attachment = dynamicDowncast<HTMLAttachmentElement>(element)) {
+        if (attachment->file()) {
+            RetainPtr<NSURL> url = [NSURL fileURLWithPath:attachment->file()->path()];
```
Probably don't need `<NSURL>`?

cdumez reviewed (approved):
LGTM

annevk committed:
Adopt dynamicDowncast<> in editing's HTMLConverter
https://bugs.webkit.org/show_bug.cgi?id=268442

Reviewed by Chris Dumez.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverterCaches::inlineStylePropertyForElement):
(stringFromCSSValue):
(HTMLConverterCaches::propertyValueForNode):
(HTMLConverterCaches::floatPropertyValueForNode):
(HTMLConverter::_blockLevelElementForNode):
(HTMLConverterCaches::colorPropertyValueForNode):
(HTMLConverter::_addTableCellForElement):
(HTMLConverter::_processElement):
(HTMLConverter::_traverseNode):
(fileWrapperForElement):
(WebCore::editingAttributedString):

Canonical link: https://commits.webkit.org/273881@main
